google-cloud-clients/google-cloud-contrib/google-cloud-nio/src/main/java/com/google/cloud/storage/contrib/nio/CloudStorageWriteFileChannel.java

```diff
@@ -0,0 +1,199 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.cloud.storage.contrib.nio;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.MappedByteBuffer;
+import java.nio.channels.FileChannel;
+import java.nio.channels.FileLock;
+import java.nio.channels.ReadableByteChannel;
+import java.nio.channels.SeekableByteChannel;
+import java.nio.channels.WritableByteChannel;
+
+class CloudStorageWriteFileChannel extends FileChannel {
+  private static final String WRITE_ONLY = "This FileChannel is write-only";
+  private SeekableByteChannel writeChannel;
+  private boolean valid = true;
+
+  CloudStorageWriteFileChannel(SeekableByteChannel writeChannel) {
+    this.writeChannel = writeChannel;
+  }
+
+  private void checkValid() throws IOException {
+    if (!valid) {
+      // These methods are only supported to be called once, because the underlying channel does not
+      // support changing the position.
+      throw new IOException(
+          "This FileChannel is no longer valid. "
+              + "A Cloud Storage FileChannel is invalidated after calling one of "
+              + "the methods FileChannel#write(ByteBuffer, long) or "
+              + "FileChannel#transferFrom(ReadableByteChannel, long, long)");
+    }
+    if (!writeChannel.isOpen()) {
+      throw new IOException("This FileChannel is closed");
+    }
+  }
+
+  @Override
+  public int read(ByteBuffer dst) throws IOException {
+    throw new UnsupportedOperationException(WRITE_ONLY);
+  }
+
+  @Override
+  public long read(ByteBuffer[] dsts, int offset, int length) throws IOException {
+    throw new UnsupportedOperationException(WRITE_ONLY);
+  }
+
+  @Override
+  public int write(ByteBuffer src) throws IOException {
+    synchronized (this) {
+      checkValid();
+      return writeChannel.write(src);
+    }
+  }
+
+  @Override
+  public long write(ByteBuffer[] srcs, int offset, int length) throws IOException {
+    synchronized (this) {
+      checkValid();
+      long res = 0L;
+      for (int i = offset; i < offset + length; i++) {
+        res += writeChannel.write(srcs[i]);
+      }
+      return res;
+    }
+  }
+
+  @Override
+  public long position() throws IOException {
+    synchronized (this) {
+      checkValid();
+      return writeChannel.position();
+    }
+  }
+
+  @Override
+  public FileChannel position(long newPosition) throws IOException {
+    synchronized (this) {
+      checkValid();
+      if (newPosition != position()) {
```

Comment:
There's a call to `checkValid()` in `position()`.  We don't need two calls.  Perhaps just use `writeChannel.position()`

Reply:
There are other places where the double check happens as well, further down.

Reply:
I removed the redundant validation checks and changed the method signatures in this class as well to synchronized.